### ISSUE TITLE
Analyzer_plugin readme unclear

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/sdk#35516]
In the following readme: https://github.com/dart-lang/sdk/tree/master/pkg/analyzer_plugin

There is this note:

> Note: The plugin support is not currently available for general use.

But this doesn't explain _why_ it is not usable yet. Is it because its version is 0.0.1-alpha.5? Or something that only Google can use as of now?

The thing is, as of now two published packages use it:
- built_value
- angular

A package of mine would massively benefit from reporting errors to the analyzer. So I'm fine with using something unstable, but this note makes it unclear if it's even doable.

**Comment by bwilkerson** [comment]
Sorry about the lack of clarity. I should probably polish the following and include it in the docs, but I can respond more quickly to your question if I don't spend too much time polishing it (and I keep hoping that the issue can be resolved soon so that the disclaimer is no longer necessary :-).

This is not the first attempt we've made to add plugin support to the analyzer, but it is the most promising so far. This approach runs plugins by loading each plugin into a separate isolate. This allows us to dynamically load and unload plugins as the user opens and closes directories, which makes the whole mechanism much more flexible than previous attempts were.

But because we cannot share memory across isolates, each plugin must perform it's own analysis of any code for which results are to be produced. That causes us to be concerned about performance. Fully analyzing all of the user's code is fairly expensive. By adding one plugin we're doing twice the analysis work. The increase in work ought to be linear in the number of plugins, but we have not yet verified that.

Even so, at some point, as a result of adding more plugins, all of the additional work being done will begin to negatively affect the user experience. We don't yet know what that point is, but we have been hoping that we could measure the impact and determine whether there is a problem that needs to be addressed before encouraging the community to write their own plugins. Unfortunately, that work hasn't been done yet, and I don't know when we'll find the time for it.

> A package of mine would massively benefit from reporting errors to the analyzer. So I'm fine with using something unstable, but this note makes it unclear if it's even doable.

It's definitely doable, as evidenced by the existing plugins :-), and there's nothing preventing you from doing so. (We don't believe in giving special privileges to some users that aren't available to everyone, even, or especially, giving those privileges to ourselves.) My (and I think _our_) only concern is to make using Dart a high quality experience for all of our users.

If you decide to create a plugin for your package, we'll be happy to answer your questions. The best place for such questions is a forum we created for people attempting to use the analyzer packages (https://groups.google.com/a/dartlang.org/forum/#!forum/analyzer-discuss).

**Comment by rrousselGit** [comment]
Thanks for the awesome explanation!

A potentially dumb question: Since the plugin and the analysis server communicate with each other; can't the analysis server send these informations to the plugin?

**Comment by bwilkerson** [comment]
> Since the plugin and the analysis server communicate with each other; can't the analysis server send these informations to the plugin?

Yes, it could. But server would need to serialize the data structures and the plugins would need to deserialize them. If there are typically a large enough number of plugins then that might prove to be a performance win, but with a small number of plugins (I think we tested with one plugin) the time to serialize and deserialize is greater than the time to re-analyze.

But the potential need to make that kind of change to the wire protocol is exactly the reason we've hesitated to promote this feature. It would be difficult to migrate existing plugins, so the fewer there are the easier it would be.
